### ISSUE TITLE
"Set internalResponse’s CSP list" never sets it on the non-internal response?

### ISSUE THREAD

**Issue by domenic** [whatwg/fetch#1002]
https://fetch.spec.whatwg.org/commit-snapshots/d5c3d8b5570074e05c86a552e8a738daa19976cd/#main-fetch

> Set internalResponse’s CSP list\. \[CSP\]

As far as I can tell, this means that when HTML calls "initialize a Document's CSP list" ([here](https://html.spec.whatwg.org/#initialise-the-document-object)) the given response's CSP list is always empty, so CSP never applies.

Probably I'm missing something?

**Comment by annevk** [comment]
The idea is that fields that are not filtered remain the same. That probably ought to be clarified. It's also not entirely clear to me anymore why we store CSP list on responses though. Is there ever a response without a `Content-Security-Policy` header that still ends up with a CSP list?

**Comment by domenic** [comment]
> The idea is that fields that are not filtered remain the same. That probably ought to be clarified.

Oh, interesting, so "Set response to the following filtered response with response as its internal response" is meant to create a sort of live two-way binding, where changes to the internal response also affect the filtered response? Very interesting. The spec text works as intended then, but yeah, it's not super-clear. Maybe I can try to submit something to clarify.

> It's also not entirely clear to me anymore why we store CSP list on responses though.

Working with a similar setup for origin policy, I can say that this setup is quite convenient. Various algorithms take a response as input and really want to look at the CSP list (or origin policy). For example most of the sub-algorithms in https://w3c.github.io/webappsec-csp/#html-integration and their call sites in HTML.

> Is there ever a response without a Content-Security-Policy header that still ends up with a CSP list?

Well, after origin policy, there will be :).

**Comment by annevk** [comment]
So that means that every policy that might end up being backed by Origin Policy would have to be on a response? E.g., we'd have to store COOP+COEP on a response? From the link you gave it seems to be primarily about initialization of the document/worker environment which will also have access to the headers of the response and the Origin Policy, if any.

I'd rather build less on top of response directly as usually this is not applicable generally (e.g., it's only for responses that end up creating globals) and it puts lots of logic in a subsystem that ultimately isn't really responsible for it.

**Comment by annevk** [comment]
cc @mikewest @andypaicu

**Comment by domenic** [comment]
> So that means that every policy that might end up being backed by Origin Policy would have to be on a response? E.g., we'd have to store COOP+COEP on a response?

I mean, kind of? An origin policy itself will be on the response, since it comes with the response. And COOP + COEP will be on the response since they are response headers.

> From the link you gave it seems to be primarily about initialization of the document/worker environment which will also have access to the headers of the response and the Origin Policy, if any.

Right, they have access to the response. And via the response they can get headers and origin policy.

> I'd rather build less on top of response directly as usually this is not applicable generally (e.g., it's only for responses that end up creating globals) and it puts lots of logic in a subsystem that ultimately isn't really responsible for it.

That seems like a large architectural change from how things currently work, and at least in Chrome, our implementation matches the current response-based spec.

**Comment by annevk** [comment]
Ah, it does make some sense for the response to carry a pointer to the origin policy in effect for it. But that doesn't really require the other things to be available as parsed entities on the response. All that could still be live in the respective policy specification. It would also have to cover how it interacts with Origin Policy, but that seems needed anyway?

**Comment by domenic** [comment]
Hmm, so I guess you are comparing:

 - Doing the parsing once and storing the parsed results on the response (e.g. https://w3c.github.io/webappsec-csp/#set-response-csp-list and then all the places in https://w3c.github.io/webappsec-csp/#html-integration that use response's CSP list), vs.
- Having each place that would use the CSP list recieve a response and parse the header there?

As I said, the former (i.e. the current strategy) seems to match Chrome's implementation more, and seems pretty nice to me. But I guess they are isomorphic; you just need to take more care to synchronize all the different parsing sites if you take the latter approach.

**Comment by annevk** [comment]
Yeah, though I mainly see it as "delaying" of the parsing of the header value. (See also the `Content-DPR` discussion for instance where I suggested it would make more sense to do the parsing where it ends up being used rather than have Fetch take care of it.) (Certainly all parsing of that header value should go through a centralized "extract X from a header list" thingy.)
